### ISSUE TITLE
Proposal: Require named arguments

### ISSUE THREAD

**Issue by JohanLarsson** [DotNetAnalyzers/StyleCopAnalyzers#2220]
Say we have a method or constructor with many parameters of the same type:

```c#
public Foo(int a, int b, int c, int d)
{
    ...
}
```

If the number of arguments are > 3? the rule should require naming them:

```c#
new Foo(a: 1, b: 2, c: 3, d: 4);
```

This comes up most often for constructors of immutable types.

**Comment by vweijsters** [comment]
👍 for me. I think this should be made `DisabledByDefault` though, as I imagine that not everybody will agree that this is a good rule.

**Comment by sharwell** [comment]
I find this a little too subjective to vote for including it here. I would prefer to see it as part of a separate analyzer project.

**Comment by sharwell** [comment]
💭 I wonder how helpful a "StyleCop Extras" project would be, where many of the subjective rules that get requested could be developed.

**Comment by ArthurHNL** [comment]
I would argue that the warning should not be fired when the argument name is equal to the name of the variable that is used as the argument:

```csharp

public void DoSomething(int a, int b, int c)
{
}

public void SomeMethod()
{
   int a = 1;
   int foo = 2;
   int d = 3;

   DoSomething(
      a, // No warning, argument name is the same as variable name.
      foo, // Warning, no argument name is provided AND the argument name is not equal to the argument name.
      c : d); // No warning, argument name is provided.
}

```
